File: C++/kth_symbol_in_grammar.cpp
// no. 779
#include <iostream>
using namespace std;

class Solution{
    public:
        int kthGrammar(int n, int k) {
            if(n == 1){
                return 0;
            }          
            int prev_k = kthGrammar(n-1, int((k+1)/2));
            if (prev_k){ // 1 -> 10
                return k % 2;
            }else{ // 0 -> 01
                return (k + 1) % 2;
            }
            
        }
};

int main(){
    Solution sol;
    cout << sol.kthGrammar(4, 3);
}

File: python/kth_symbol_in_grammar.py
# no. 779
class Solution:
    def kthGrammar(self, n: int, k:int) -> int:
        mapping = {0: [0, 1], 1: [1, 0]}
        if n == 1:
            return 0
        prev_k = self.kthGrammar(n-1, int((k+1)/2))
        if k%2:
            m = 0
        else:
            m = 1
        return mapping[prev_k][m]

if __name__ == '__main__':
    solution = Solution()
    print(solution.kthGrammar(4, 3))

